service/src/main/java/bio/terra/workspace/service/resource/controlled/cloud/azure/storageContainer/CreateAzureStorageContainerStep.java

```diff
@@ -0,0 +1,142 @@
+
+  @Override
+  public StepResult doStep(FlightContext context) throws InterruptedException, RetryException {
+    final AzureCloudContext azureCloudContext =
+        context
+            .getWorkingMap()
+            .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+    final StorageManager storageManager = crlService.getStorageManager(azureCloudContext, azureConfig);
+
+    try {
+      storageManager
+          .blobContainers()
+          .defineContainer(resource.getStorageContainerName())
+          .withExistingStorageAccount(azureCloudContext.getAzureResourceGroupId(), resource.getStorageAccountName())
+          .withPublicAccess(PublicAccess.NONE)
+          .withMetadata("workspaceId", resource.getWorkspaceId().toString())
+          .withMetadata("resourceId", resource.getResourceId().toString()).
+              create(
+                Defaults.buildContext(
+                  CreateStorageContainerRequestData.builder()
+                      .setName(resource.getStorageContainerName())
+                      .setStorageAccountName(resource.getStorageAccountName())
+                      .setResourceGroupName(azureCloudContext.getAzureResourceGroupId())
+                      .build()));
+
+    } catch (ManagementException e) {
+      logger.error(
+          "Failed to create the Azure storage container '{}' with storage account with the name '{}'. Error Code: {}",
+          resource.getStorageContainerName(),
+          resource.getStorageAccountName(),
+          e.getValue().getCode(),
+          e);
+
+      return new StepResult(StepStatus.STEP_RESULT_FAILURE_RETRY, e);
+    }
+
+    return StepResult.getStepResultSuccess();
+  }
+
+  /**
+   * Deletes the storage container if the container is available. If the storage container is available
+   * and deletes fails, the failure is considered fatal and must looked into it.
+   *
+   * @param context
+   * @return Step result.
+   * @throws InterruptedException
+   */
+  @Override
+  public StepResult undoStep(FlightContext context) throws InterruptedException {
+    final AzureCloudContext azureCloudContext =
+            context
+                    .getWorkingMap()
+                    .get(ControlledResourceKeys.AZURE_CLOUD_CONTEXT, AzureCloudContext.class);
+    final StorageManager storageManager = crlService.getStorageManager(azureCloudContext, azureConfig);
+
+    try {
+      storageManager
+              .storageAccounts()
+              .getByResourceGroup(
+                      azureCloudContext.getAzureResourceGroupId(), resource.getStorageAccountName());
+    } catch (ManagementException ex) {
+      logger.warn(
+              "Deletion of the storage container is not required. Parent storage account does not exist. {}",
+              resource.getStorageAccountName());
+      return StepResult.getStepResultSuccess();
+    }
+    final PagedIterable<ListContainerItemInner> existingContainers = storageManager.blobContainers().list(
```

Comment:
❓ Would it make sense to use [`blobContainers.get()` ](https://docs.microsoft.com/en-us/java/api/com.azure.resourcemanager.storage.models.blobcontainers.get?view=azure-java-stable#com-azure-resourcemanager-storage-models-blobcontainers-get(java-lang-string-java-lang-string-java-lang-string)) to see if the container exists, rather than iterating over the containers in the account?

Reply:
I shied away from the `get` method because I think it throws an exception if the container doesn't exist. As far as I can tell, the `list` method is the closest to `exists`.

I can call `get` and handle the exception if folks think that is clearer (I will also double-check that it throws an exception if the container DNE-- the documentation does not say).